File: Graph/Network_Delay_Time.py
#problem Link :- https://leetcode.com/problems/network-delay-time/

"""There are N network nodes, labelled 1 to N.Given times, a list of travel times as 
directed edges times[i] = (u, v, w), where u is the source node, v is the target node, 
and w is the time it takes for a signal to travel from source to target.Now, we send a 
signal from a certain node K. How long will it take for all nodes to receive the signal? 
If it is impossible, return -1."""

class Solution(object):
    from collections import heapq
    def networkDelayTime(self, times, N, K):
        from collections import defaultdict
        graph = defaultdict(list)
        for u,v,w in times:
            graph[u].append((v,w))
        
        pq = [(0,K)]
        dist = {}
        while pq:
            d,node = heappop(pq)
            if node in dist:
                continue
            dist[node] = d
            for nei,d2 in graph[node]:
                if nei not in dist:
                    heapq.heappush(pq,(d+d2,nei))
        return max(dist.values()) if len(dist) == N else -1
            

File: Graph/NetworkDelayTime.cpp
//Problem Link:- https://leetcode.com/problems/network-delay-time/

#include<bits/stdc++.h>
using namespace std;

class Solution {
    private:
        vector<pair<int,int>> graph[105];
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        for(int i=0;i<times.size();i++){
            int u = times[i][0];
            int v = times[i][1];
            int t = times[i][2];
            graph[u].push_back({v,t});
        }
        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> pq;
        pq.push({0,k});
        vector<int> dist(n+1,INT_MAX);
        dist[k] = 0;
        while (!pq.empty()){
            auto node = pq.top();
            pq.pop();
            int d = node.first;
            int u = node.second;
            for(auto nie : graph[u]){
                int child = nie.first;
                int w = nie.second;
                if (dist[child] > dist[u] + w){
                    dist[child] = dist[u] + w;
                    pq.push({dist[child],child});
                }
            }
        }
        int res = *max_element(dist.begin()+1,dist.end());
        return (res == INT_MAX ? -1 : res);
    }
};